Add type annotations to Trinity code

### What is wrong?

Need to add type annotations before Trinity code gets any larger.

### How can it be fixed

Add them

Ok, turns out I'm running into unexpected challenges here. Having no experience with the type annotation support in Python I assumed things to behave differently.

This works as expected

```Python
def yay(val: bool):
  print(val)

yay(5)
```

Gives me:

```shell
$ mypy test.py 
test.py:11: error: Argument 1 to "yay" has incompatible type "int"; expected "bool"
```

But for instance, this doesn't raise any type error even though I assume it should (calling with `bool` where `str` is expected)

```Python
class Foo:
  def __init__(self):
    self.bar(True)

  def bar(self, val: str):
    print(val)
```

Same here, this doesn't raise a type error even though I would expect it to raise one since I'm assigning `bool` where a `str` is expected.

```Python
class Foo:
  bar:str = None
  def __init__(self):
    self.bar = True

```

Sorry, I don't mean to turn this into a free Python lesson for me but I'm not sure if I could really tackle this  task if the tooling doesn't give me the feedback I need. Chances are I'm completely approaching it wrong.

So, if you have pointers for me that would be gladly appreciated ;)

--- a/trinity/utils/chains.py
+++ b/trinity/utils/chains.py
@@ -23,11 +23,13 @@ from .xdg import (
     get_xdg_trinity_root,
 )
 
+from typing import Union
+
 
 #
 # Filesystem path utils
 #
-def get_default_data_dir(chain_identifier):
+def get_default_data_dir(chain_identifier: str) -> str:
     """
     Returns the base directory path where data for a given chain will be stored.
     """
@@ -40,7 +42,7 @@ def get_default_data_dir(chain_identifier):
 DATABASE_DIR_NAME = 'chain'
 
 
-def get_database_dir(chain_identifier, data_dir=None):
+def get_database_dir(chain_identifier: str, data_dir: str=None) -> str:
     """
     Returns the directory path where chain data will be stored.
     """
@@ -52,7 +54,7 @@ def get_database_dir(chain_identifier, data_dir=None):
 NODEKEY_FILENAME = 'nodekey'
 
 
-def get_nodekey_path(chain_identifier, data_dir=None):
+def get_nodekey_path(chain_identifier: str, data_dir: str=None) -> str:
     """
     Returns the path to the private key used for devp2p connections.
     """
@@ -67,7 +69,7 @@ def get_nodekey_path(chain_identifier, data_dir=None):
 DATABASE_SOCKET_FILENAME = 'db.ipc'
 
 
-def get_database_socket_path(chain_identifier, data_dir=None):
+def get_database_socket_path(chain_identifier: str, data_dir: str=None) -> str:
     """
     Returns the path to the private key used for devp2p connections.
     """
@@ -82,7 +84,7 @@ def get_database_socket_path(chain_identifier, data_dir=None):
 #
 # Nodekey loading
 #
-def load_nodekey(nodekey_path):
+def load_nodekey(nodekey_path: str) -> PrivateKey:
     with open(nodekey_path, 'rb') as nodekey_file:
         nodekey_raw = nodekey_file.read()
     nodekey = keys.PrivateKey(nodekey_raw)
@@ -106,11 +108,11 @@ class ChainConfig:
     _network_id = None
 
     def __init__(self,
-                 chain_identifier,
-                 data_dir=None,
-                 nodekey_path=None,
-                 nodekey=None,
-                 network_id=None):
+                 chain_identifier: str,
+                 data_dir: str=None,
+                 nodekey_path: str=None,
+                 nodekey: PrivateKey=None,
+                 network_id: int=None) -> None:
         # validation
         if nodekey is not None and nodekey_path is not None:
             raise ValueError("It is invalid to provide both a `nodekey` and a `nodekey_path`")
@@ -127,7 +129,7 @@ class ChainConfig:
             self.nodekey = nodekey
 
     @property
-    def data_dir(self):
+    def data_dir(self) -> str:
         """
         The data_dir is the base directory that all chain specific information
         for a given chain is stored.  All other chain directories are by
@@ -139,19 +141,19 @@ class ChainConfig:
             return self._data_dir
 
     @data_dir.setter
-    def data_dir(self, value):
+    def data_dir(self, value: str) -> None:
         self._data_dir = os.path.abspath(value)
 
     @property
-    def database_dir(self):
+    def database_dir(self) -> str:
         return get_database_dir(self.chain_identifier, self.data_dir)
 
     @property
-    def database_ipc_path(self):
+    def database_ipc_path(self) -> str:
         return get_database_socket_path(self.chain_identifier, self.data_dir)
 
     @property
-    def nodekey_path(self):
+    def nodekey_path(self) -> str:
         if self._nodekey_path is None:
             if self._nodekey is not None:
                 return None
@@ -161,11 +163,11 @@ class ChainConfig:
             return self._nodekey_path
 
     @nodekey_path.setter
-    def nodekey_path(self, value):
+    def nodekey_path(self, value: str) -> None:
         self._nodekey_path = os.path.abspath(value)
 
     @property
-    def nodekey(self):
+    def nodekey(self) -> PrivateKey:
         if self._nodekey is None:
             try:
                 return load_nodekey(self.nodekey_path)
@@ -180,7 +182,7 @@ class ChainConfig:
             return self._nodekey
 
     @nodekey.setter
-    def nodekey(self, value):
+    def nodekey(self, value: Union[bytes, PrivateKey]) -> None:
         if isinstance(value, bytes):
             self._nodekey = keys.PrivateKey(value)
         elif isinstance(value, PrivateKey):
@@ -192,7 +194,7 @@ class ChainConfig:
             )
 
     @property
-    def network_id(self):
+    def network_id(self) -> int:
         if self._network_id is not None:
             return self._network_id
 

--- a/trinity/utils/db.py
+++ b/trinity/utils/db.py
@@ -1,35 +1,38 @@
-class MemoryDB:
-    kv_store = None
+from typing import Dict
 
-    def __init__(self, kv_store=None):
+
+class MemoryDB:
+    kv_store = None  # type: Dict[bytes, bytes]
+
+    def __init__(self, kv_store: Dict[bytes, bytes]=None) -> None:
         if kv_store is None:
             self.kv_store = {}
         else:
             self.kv_store = kv_store
 
-    def get(self, key):
+    def get(self, key: bytes) -> bytes:
         return self.kv_store[key]
 
-    def set(self, key, value):
+    def set(self, key: bytes, value: bytes) -> None:
         self.kv_store[key] = value
 
-    def exists(self, key):
+    def exists(self, key: bytes) -> bool:
         return key in self.kv_store
 
-    def delete(self, key):
+    def delete(self, key: bytes) -> None:
         del self.kv_store[key]
 
     #
     # Dictionary API
     #
-    def __getitem__(self, key):
+    def __getitem__(self, key: bytes) -> bytes:
         return self.get(key)
 
-    def __setitem__(self, key, value):
+    def __setitem__(self, key: bytes, value: bytes) -> None:
         return self.set(key, value)
 
-    def __delitem__(self, key):
+    def __delitem__(self, key: bytes) -> None:
         return self.delete(key)
 
-    def __contains__(self, key):
+    def __contains__(self, key: bytes) -> bool:
         return self.exists(key)

--- a/trinity/utils/filesystem.py
+++ b/trinity/utils/filesystem.py
@@ -1,7 +1,7 @@
 import os
 
 
-def is_same_path(p1, p2):
+def is_same_path(p1: str, p2: str) -> bool:
     n_p1 = os.path.abspath(os.path.expanduser(p1))
     n_p2 = os.path.abspath(os.path.expanduser(p2))
 
@@ -11,7 +11,7 @@ def is_same_path(p1, p2):
         return n_p1 == n_p2
 
 
-def is_under_path(base_path, path, strict=True):
+def is_under_path(base_path: str, path: str, strict: bool=True) -> bool:
     if is_same_path(base_path, path):
         if strict:
             return False

--- a/trinity/utils/ipc.py
+++ b/trinity/utils/ipc.py
@@ -1,9 +1,11 @@
 import os
 import signal
 import time
+from multiprocessing import Process
+from logging import Logger
 
 
-def wait_for_ipc(ipc_path, timeout=1):
+def wait_for_ipc(ipc_path: str, timeout: int=1) -> None:
     start_at = time.time()
     while time.time() - start_at < timeout:
         if os.path.exists(ipc_path):
@@ -11,7 +13,10 @@ def wait_for_ipc(ipc_path, timeout=1):
         time.sleep(0.05)
 
 
-def kill_process_gracefully(process, logger=None, SIGINT_timeout=5, SIGTERM_timeout=3):
+def kill_process_gracefully(process: Process,
+                            logger: Logger=None,
+                            SIGINT_timeout: int=5,
+                            SIGTERM_timeout: int=3) -> None:
     try:
         if not process.is_alive():
             return

--- a/trinity/utils/logging.py
+++ b/trinity/utils/logging.py
@@ -1,12 +1,15 @@
 import functools
 import logging
-from logging import handlers
+from logging import handlers, Logger
+from multiprocessing import Queue
 import sys
 
 from cytoolz import dissoc
 
+from typing import Tuple, Callable
 
-def setup_trinity_logging(level):
+
+def setup_trinity_logging(level: int) -> Tuple[Logger, Queue, handlers.QueueListener]:
     from .mp import ctx
 
     log_queue = ctx.Queue()
@@ -29,7 +32,7 @@ def setup_trinity_logging(level):
     return logger, log_queue, listener
 
 
-def setup_queue_logging(log_queue, level):
+def setup_queue_logging(log_queue: Queue, level: int) -> None:
     queue_handler = handlers.QueueHandler(log_queue)
     logging.basicConfig(
         level=level,
@@ -40,7 +43,7 @@ def setup_queue_logging(log_queue, level):
     logger.debug('Logging initialized')
 
 
-def with_queued_logging(fn):
+def with_queued_logging(fn: Callable) -> Callable:
     @functools.wraps(fn)
     def inner(*args, **kwargs):
         print(args, kwargs)

--- a/trinity/utils/xdg.py
+++ b/trinity/utils/xdg.py
@@ -21,7 +21,7 @@ XDG_DATA_HOME = os.environ.get(
 )
 
 
-def get_xdg_trinity_root():
+def get_xdg_trinity_root() -> str:
     """
     Returns the base directory under which trinity will store all data.
     """
@@ -31,7 +31,7 @@ def get_xdg_trinity_root():
     )
 
 
-def is_under_xdg_trinity_root(path):
+def is_under_xdg_trinity_root(path: str) -> bool:
     return is_under_path(
         get_xdg_trinity_root(),
         path,

